alternator/ttl.cc

```diff
@@ -548,13 +550,26 @@ static future<> scan_table_ranges(
             co_return;
         }
         auto units = co_await get_units(page_sem, 1);
-        // We don't to limit page size in number of rows because there is a
-        // builtin limit of the page's size in bytes. Setting this limit to 1
-        // is useful for debugging the paging code with moderate-size data.
+        // We don't need to limit page size in number of rows because there is
+        // a builtin limit of the page's size in bytes. Setting this limit to
+        // 1 is useful for debugging the paging code with moderate-size data.
         uint32_t limit = std::numeric_limits<uint32_t>::max();
-        // FIXME: which timeout?
-        // FIXME: if read times out, need to retry it.
-        std::unique_ptr<cql3::result_set> rs = co_await p->fetch_page(limit, gc_clock::now(), executor::default_timeout());
+        // Read a page, and if that times out, try again after a small sleep.
+        // If we didn't catch the timeout exception, it would cause the scan
+        // be aborted and only be restarted at the next scanning period.
+        std::unique_ptr<cql3::result_set> rs;
+        for (;;) {
+            try {
+                // FIXME: which timeout?
+                rs = co_await p->fetch_page(limit, gc_clock::now(), executor::default_timeout());
+                break;
+            } catch(exceptions::read_timeout_exception&) {
+                tlogger.warn("expiration scanner read timed out, will retry: {}",
+                    std::current_exception());
+            }
+            // If we didn't break out of this loop, add a minimal sleep
+            co_await seastar::sleep(1s);
```

Comment:
I don't know how it's typically done in scylla, but in general wouldn't be better to have: 
- limit of retries
- some more advanced backoff (like linear/exponential)

the latter maybe of less importance as I understand there not that many such scans running (at most one?)

Reply:
Yes, I don't think we need exponential backoff because there's indeed just one of these threads. I guess limiting the number of retries makes sense in case some sort of code bug in our code causes a reproducible hang (timeout) in the request, and we can get stuck in the same place forever - it may indeed better to give up after some time and start the scan from scratch at a different position. I'll think about it.

Reply:
I was thinking more about retry limit, perhaps it's ok, and what we need to make sure it's that at most one TTL "job" is active, i.e. that the next one cancels the pending/stuck one with some warn log. Maybe it's already the case?